src/stream/src/executor/backfill/snapshot_backfill.rs

```diff
@@ -0,0 +1,494 @@
+use futures::future::Either;
+use futures::{pin_mut, Stream, TryStreamExt};
+use itertools::Itertools;
+use risingwave_common::array::{Op, StreamChunk};
+use risingwave_common::catalog::TableId;
+use risingwave_common::row::OwnedRow;
+use risingwave_common::util::chunk_coalesce::DataChunkBuilder;
+use risingwave_hummock_sdk::HummockReadEpoch;
+use risingwave_storage::error::StorageResult;
+use risingwave_storage::table::batch_table::storage_table::StorageTable;
+use risingwave_storage::table::ChangeLogRow;
+use risingwave_storage::StateStore;
+use tokio::select;
+use tokio::sync::mpsc::UnboundedReceiver;
+
+use crate::executor::backfill::utils::{create_builder, mapping_chunk};
+use crate::executor::prelude::{try_stream, StreamExt};
+use crate::executor::{
+    expect_first_barrier, ActorContextRef, BackfillExecutor, Barrier, BoxedMessageStream, Execute,
+    Executor, Message, Mutation, StreamExecutorError, StreamExecutorResult,
+};
+use crate::task::CreateMviewProgress;
+
+pub struct SnapshotBackfillExecutor<S: StateStore> {
+    /// Upstream table
+    upstream_table: StorageTable<S>,
+
+    /// Upstream with the same schema with the upstream table.
+    upstream: Executor,
+
+    /// The column indices need to be forwarded to the downstream from the upstream and table scan.
+    output_indices: Vec<usize>,
+
+    progress: CreateMviewProgress,
+
+    chunk_size: usize,
+    rate_limit: Option<usize>,
+
+    barrier_rx: UnboundedReceiver<Barrier>,
+}
+
+impl<S: StateStore> SnapshotBackfillExecutor<S> {
+    #[expect(clippy::too_many_arguments)]
+    pub fn new(
+        upstream_table: StorageTable<S>,
+        upstream: Executor,
+        output_indices: Vec<usize>,
+        _actor_ctx: ActorContextRef,
+        progress: CreateMviewProgress,
+        chunk_size: usize,
+        rate_limit: Option<usize>,
+        barrier_rx: UnboundedReceiver<Barrier>,
+    ) -> Self {
+        if let Some(rate_limit) = rate_limit {
+            debug!(
+                rate_limit,
+                "create snapshot backfill executor with rate limit"
+            );
+        }
+        Self {
+            upstream_table,
+            upstream,
+            output_indices,
+            progress,
+            chunk_size,
+            rate_limit,
+            barrier_rx,
+        }
+    }
+
+    #[try_stream(ok = Message, error = StreamExecutorError)]
+    async fn execute_inner(mut self) {
+        debug!("snapshot backfill executor start");
+        let mut upstream = self.upstream.execute();
+        let upstream_table_id = self.upstream_table.table_id();
+        let first_barrier = expect_first_barrier(&mut upstream).await?;
+        debug!(epoch = ?first_barrier.epoch, "get first upstream barrier");
+        let first_recv_barrier = receive_next_barrier(&mut self.barrier_rx).await?;
+        debug!(epoch = ?first_recv_barrier.epoch, "get first inject barrier");
+        let should_backfill = first_barrier.epoch != first_recv_barrier.epoch;
+
```

Comment:
I suppose the initial iteration of snapshot backfill is entirely in-memory?

Reply:
Yes, the current snapshot backfill is not recoverable. If recovery happens before marked as created, the job will be clean and failed.